tests/tests/util/contracts.ts

```diff
@@ -0,0 +1,149 @@
+import { ApiPromise } from "@polkadot/api";
+import { TransactionReceipt } from "web3-core";
+import { AbiItem } from "web3-utils";
+import { Contract } from "web3-eth-contract";
+import { createAndFinalizeBlock } from "./polkadotApiRequests";
+import { Context } from "./testWithMoonbeam";
+import solc from "solc";
+import Web3 from "web3";
+import fs from "fs";
+
+import { GENESIS_ACCOUNT, GENESIS_ACCOUNT_PRIVATE_KEY } from "../constants";
+import { customRequest } from ".";
+import { contractSources } from "../constants/contractSources";
+
+export function compileSolidity(contractContent: string, contractName: string = "Test"): Compiled {
+  let result = JSON.parse(
+    solc.compile(
+      JSON.stringify({
+        language: "Solidity",
+        sources: {
+          "main.sol": {
+            content: contractContent,
+          },
+        },
+        settings: {
+          outputSelection: {
+            "*": {
+              "*": ["*"],
+            },
+          },
+        },
+      })
+    )
+  );
+
+  const contract = result.contracts["main.sol"][contractName];
+  return {
+    byteCode: "0x" + contract.evm.bytecode.object,
+    contract,
+  };
+}
+export interface Compiled {
+  byteCode: string;
+  contract: any;
+}
+
+export async function getCompiled(name: string): Promise<Compiled> {
```

Comment:
I'm not sure this is thread-safe. But it shouldn't be an issue if we think of adding the compiled contract to the repo after

Reply:
why wouldn't it be thread safe?

Reply:
Because 2 ore more tests can execute that function at the same time (in different threads). In that case, each one will see that the contract doesn't exist and will compile it and as a result will try to write to that file at the same time.
I'm fairly convinced that writeFile is not atomic and so you could end up with a file containing chunks of data from one thread and chunks from the others

Reply:
ha I see. So what is the solution? Hopefully the compilation will only happen once when the contract is added

Reply:
I think we should not compile on the fly but precompile using the mocha fixtures in that case. It runs only once before all the tests.

Reply:
OK but then it means that the fixture must know of all contracts used in the test suite

Reply:
also if we want to do that we can just do it in DescribeWithMoonbeam wihtout using fixture.

Reply:
No because the describeWithMoonbeam gets executed by multiple threads and you will have the same issue of race condition.
In the fixture, I'd process all the contracts in the factory (They should already be done if the json files are there so it should not have any cost except checking if the file exists)

Reply:
ha ok no nevermind we do need fixture

Reply:
Unfortunately fixtures are only in js :/

Reply:
No, it works in TS too, I've used them when I did the `--parellel` work

Reply:
Do you have an example? is it `--require fixture.mts` ? When I did this it said I couldnt use `import` outside of a module
In the meantime Im just calling `ts-node fixture.ts`

Reply:
`-r contract-fixture.ts`

Reply:
let's get on a call because Im getiing errors:
`✖ ERROR: Error: Cannot find module './tests/constants/contractSources'`